src/Conversion/ONNXToKrnl/Math/MatMulInteger.cpp

```diff
@@ -0,0 +1,112 @@
+//
+// Copyright 2019-2023 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX MatMulInteger Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/Krnl/DialectBuilder.hpp"
+#include "src/Dialect/Krnl/KrnlHelper.hpp"
+#include "src/Dialect/Mlir/DialectBuilder.hpp"
+#include "src/Dialect/Mlir/IndexExpr.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+struct ONNXMatMulIntegerOpLowering
+    : public OpConversionPattern<ONNXMatMulIntegerOp> {
+public:
+  ONNXMatMulIntegerOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : OpConversionPattern(typeConverter, ctx) {}
+
+  LogicalResult matchAndRewrite(ONNXMatMulIntegerOp mmiOp,
+      ONNXMatMulIntegerOpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const final {
+    using LocalDialectBuilder = MultiDialectBuilder<KrnlBuilder,
+        IndexExprBuilderForKrnl, MathBuilder, MemRefBuilder, OnnxBuilder>;
+    Operation *op = mmiOp.getOperation();
+    Location loc = ONNXLoc<ONNXMatMulIntegerOp>(op);
+    LocalDialectBuilder create(rewriter, loc);
+
+    ValueRange operands = adaptor.getOperands();
+    Value A = adaptor.getA();
+    Value B = adaptor.getB();
+    Value aZeroPoint = mmiOp.getAZeroPoint(); // Optional input.
+    Value bZeroPoint = mmiOp.getBZeroPoint(); // Optional input.
+
+    if (!isNoneValue(aZeroPoint)) {
+      ShapedType stype = aZeroPoint.getType().dyn_cast<ShapedType>();
+      if ((stype.getRank() > 1) ||
+          (stype.getRank() == 1 && (stype.getShape()[0] != 1))) {
+        return rewriter.notifyMatchFailure(op, [&](mlir::Diagnostic &diag) {
+          diag << "Does not support non-scalar for a_zero_point";
+        });
+      }
+    }
+
+    // Common types.
+    auto resMemRefType = dyn_cast<MemRefType>(
+        typeConverter->convertType(mmiOp.getResult().getType()));
+    Type resElementType = resMemRefType.getElementType();
+
+    // Get shape.
+    ONNXMatMulIntegerOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    Type i32Ty = rewriter.getIntegerType(32);
+    Type ui32Ty = rewriter.getIntegerType(32, /*isSigned=*/false);
+    assert(resElementType == i32Ty && "Output type is not i32");
+
+    Value AUInt32, BUInt32;
+    // Use i32 for onnx.Sub because onnx.Sub does not support ui32.
+    // It's because `arith` dialect does not have `subui` though it has `addui`
+    // and `mului`.
+    if (!isNoneValue(aZeroPoint)) {
+      Value AInt32 = create.onnx.cast(A, i32Ty);
+      Value aZeroPointInt32 = create.onnx.cast(aZeroPoint, i32Ty);
+      // Note: `sub` is not true if aZeroPoint shape is [M], M != 1,
+      // because [MxK] - [M] cannot be expressed by `sub`.
```

Comment:
> [MxK] - [M] cannot be expressed by `sub`
we could unsqueeze aZeroPoint like we do in MatMulInteger const prop: https://github.com/onnx/onnx-mlir/blob/main/src/Transform/ONNX/ConstProp.cpp#L438-L443

Reply:
I updated this to handle this case. Do you know if we have this situation: A is [D1, D2, M, K] and aZeroPoint is [D1, D2, M]? I am not sure but handle such a situation too? Thanks!

Reply:
The spec says:
> If the input is N-D tensor with shape [D1, D2, M, K] then zero point tensor may have shape [D1, D2, M, 1].

In the constant propagation implementation I chose that to mean that a zero point tensor with rank > 1 must broadcast to the corresponding multidimensional matrix input shape. See https://github.com/onnx/onnx-mlir/blob/main/src/Transform/ONNX/ConstProp.cpp#L446-L450

Reply:
[D1, D2, M, K] and [D1, D2, M, 1] is correct for broadcasting, so no problem to do subtraction. What I meant is [D1, D2, M, K] and [D1, D2, M], where we should unsqueeze. I seems the spec does not mention that case, but only ([M, K], [M]) and ([D1, D2, M, K], [D1, D2, M, 1]).

Reply:
what I meant is that I interpret the specification
> If the input is 2D of shape [M, K] then zero point tensor may be an M element vector [zp_1, zp_2, ..., zp_M]. If the input is N-D tensor with shape [D1, D2, M, K] then zero point tensor may have shape [D1, D2, M, 1].
to mean that a zero point with rank > 1 must broadcast to the N-D tensor shape as in the example [D1, D2, M, K] and [D1, D2, M, 1], so we should only unsqueeze zero point if it is has rank 1 as in the the example [M, K] and [zp_1, zp_2, ..., zp_M]

I hope that's clear from the implementation, asserts, and comments in ConstProp.cpp - the asserts are meant to alert us to any model that disagrees with this interpretation of the spec

maybe we can factor out the logic that decides how to unsqueeze a_zero_point into a function we can share between constant propagation and lowering

Reply:
Totally agree. It would be a good idea to put these constraints to the verifier for MatMulInteger so that it would be valid for both ConstProp.cpp and lowering.

Reply:
In bumpy.matmul:
```
Matrix product of two arrays.

The behavior depends on the arguments in the following way.

- If both arguments are 2-D they are multiplied like conventional matrices.
- If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.
- If the first argument is 1-D, it is promoted to a matrix by prepending a 1 to its dimensions. After matrix multiplication the prepended 1 is removed.
- If the second argument is 1-D, it is promoted to a matrix by appending a 1 to its dimensions. After matrix multiplication the appended 1 is removed.

```

Reply:
@chentong319 I believe all of that is dealt with in `onnx.matmul()`

matching the zero point shapes in MatMulInteger is separate issue, I believe